rosbag2_storage/src/rosbag2_storage/qos.cpp

```diff
@@ -63,45 +131,118 @@ bool convert<rmw_time_t>::decode(const Node & node, rmw_time_t & time)
   return true;
 }
 
-Node convert<rosbag2_transport::Rosbag2QoS>::encode(const rosbag2_transport::Rosbag2QoS & qos)
+Node convert<rosbag2_storage::Rosbag2QoS>::encode(
+  const rosbag2_storage::Rosbag2QoS & qos)
 {
   const auto & p = qos.get_rmw_qos_profile();
   Node node;
-  node["history"] = static_cast<int>(p.history);
+  node["history"] = convert<rmw_qos_history_policy_t>::encode(p.history);
   node["depth"] = p.depth;
-  node["reliability"] = static_cast<int>(p.reliability);
-  node["durability"] = static_cast<int>(p.durability);
+  node["reliability"] = convert<rmw_qos_reliability_policy_t>::encode(p.reliability);
+  node["durability"] = convert<rmw_qos_durability_policy_t>::encode(p.durability);
   node["deadline"] = p.deadline;
   node["lifespan"] = p.lifespan;
-  node["liveliness"] = static_cast<int>(p.liveliness);
+  node["liveliness"] = convert<rmw_qos_liveliness_policy_t>::encode(p.liveliness);
   node["liveliness_lease_duration"] = p.liveliness_lease_duration;
   node["avoid_ros_namespace_conventions"] = p.avoid_ros_namespace_conventions;
   return node;
 }
 
-bool convert<rosbag2_transport::Rosbag2QoS>::decode(
-  const Node & node, rosbag2_transport::Rosbag2QoS & qos)
+bool convert<rosbag2_storage::Rosbag2QoS>::decode(
+  const Node & node, rosbag2_storage::Rosbag2QoS & qos, int version)
 {
-  auto history = static_cast<rmw_qos_history_policy_t>(node["history"].as<int>());
-  auto reliability = static_cast<rmw_qos_reliability_policy_t>(node["reliability"].as<int>());
-  auto durability = static_cast<rmw_qos_durability_policy_t>(node["durability"].as<int>());
-  auto liveliness = static_cast<rmw_qos_liveliness_policy_t>(node["liveliness"].as<int>());
+  rmw_qos_history_policy_t history;
+  rmw_qos_reliability_policy_t reliability;
+  rmw_qos_durability_policy_t durability;
+  rmw_qos_liveliness_policy_t liveliness;
+
+  if (version <= 8) {
+    history = static_cast<rmw_qos_history_policy_t>(node["history"].as<int>());
+    reliability = static_cast<rmw_qos_reliability_policy_t>(node["reliability"].as<int>());
+    durability = static_cast<rmw_qos_durability_policy_t>(node["durability"].as<int>());
+    liveliness = static_cast<rmw_qos_liveliness_policy_t>(node["liveliness"].as<int>());
+  } else {
+    history = node["history"].as<rmw_qos_history_policy_t>();
+    reliability = node["reliability"].as<rmw_qos_reliability_policy_t>();
+    durability = node["durability"].as<rmw_qos_durability_policy_t>();
+    liveliness = node["liveliness"].as<rmw_qos_liveliness_policy_t>();
+  }
+
+  switch (history) {
+    case RMW_QOS_POLICY_HISTORY_KEEP_LAST:
+      qos.keep_last(node["depth"].as<int>());
+      break;
+    case RMW_QOS_POLICY_HISTORY_KEEP_ALL:
+      qos.keep_all();
+      break;
+    default:
+      qos.history(history);
+  }
```

Comment:
```suggestion
  qos.history(history);
```
It looks like the entire switch-case could be replaced by the one method call.

Reply:
If I keep the old code:
```cpp
qos
  .keep_last(node["depth"].as<int>())
  .history(history)
```
I get the following error:
```shell
FAIL: test_info.TestRos2BagInfo.test_info_with_topic_name_option
--------------------------------------------------------------------------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/roxy/ros2_rolling/src/ros2/rosbag2/ros2bag/test/test_info.py", line 68, in test_info_with_topic_name_option
    assert normalize_lineseps(bag_command.output) == EXPECTED_OUTPUT, \
AssertionError: ros2bag CLI did not produce the expected output
assert '[WARN] [1698...ts\n/rosout\n' == '/parameter_events\n/rosout\n'
  + [WARN] [1698932598.894371851] [rclcpp]: A zero depth with KEEP_LAST doesn't make sense; no data could be stored.This will be interpreted as SYSTEM_DEFAULT
```
that looking at `rclcpp` source code, happens when you have a `depth=0`. See https://github.com/ros2/rclcpp/pull/2048.

Reply:
Nitpick. I think that if-else for this purposrse would be easy to understand rationale for that.
e.g.
```
  if (history == RMW_QOS_POLICY_HISTORY_KEEP_LAST) {
    qos.keep_last(node["depth"].as<int>());
  } else {
    qos.history(history);
  }
```